### ISSUE TITLE
`collect_async` is blocking

### ISSUE THREAD

**Issue by winstxnhdw** [pola-rs/polars#18718]
### Checks

- [X] I have checked that this issue has not already been reported.
- [X] I have confirmed this bug exists on the [latest version](https://pypi.org/project/polars/) of Polars.

### Reproducible example

```python
from typing import TypedDict, Awaitable, Iterable
from polars import DataFrame, LazyFrame


class Word(TypedDict):
    id: int
    text: str
    start_time: float
    end_time: float

class Speaker(TypedDict):
    speaker: str
    start_time: float
    end_time: float


def get_intersection(transcription: Iterable[Word], diarisation: Iterable[Speaker]) -> Awaitable[DataFrame]:
    intersection_expression = min_horizontal(
        'end_time',
        'end_time_right',
    ) - max_horizontal(
        'start_time',
        'start_time_right',
    )

    return (
        LazyFrame(transcription)
        .join(LazyFrame(diarisation), how='cross')
        .with_columns(intersection_expression.alias('intersection'))
        .collect_async()
    )
```

### Issue description

Most of the CPU-bound activities happen in Rust, when the GIL is dropped. Ideally, `collect_async` should take advantage of this for `polars` to maximise CPU usage. As of right now, `collect_async` will block the main event loop and stop your server from handling any more requests until the `DataFrame` is created.

Looking at the source code for how `collect_async` is implemented, I realised that `polars` does not spawn a new thread to run the collection in. The person who implemented this has effectively created a useless async wrapper over `collect` that has none of the benefits of async/await.

### Expected behavior

`collect_async` should not block the main event loop and act as an actual async function that will allow Python to perform context switching and process other tasks that drop the GIL concurrently.

### Installed versions

<details>

```
Polars:              1.7.0
Index type:          UInt32
Platform:            Linux-5.15.153.1-microsoft-standard-WSL2-x86_64-with-glibc2.40
Python:              3.12.5 (main, Aug  9 2024, 08:20:41) [GCC 14.2.1 20240805]

----Optional dependencies----
adbc_driver_manager  <not installed>
altair               <not installed>
cloudpickle          <not installed>
connectorx           <not installed>
deltalake            <not installed>
fastexcel            <not installed>
fsspec               2024.9.0
gevent               <not installed>
great_tables         <not installed>
matplotlib           3.9.2
nest_asyncio         <not installed>
numpy                1.26.4
openpyxl             <not installed>
pandas               2.2.2
pyarrow              16.1.0
pydantic             2.9.1
pyiceberg            <not installed>
sqlalchemy           2.0.34
torch                2.3.1+cu121
xlsx2csv             <not installed>
xlsxwriter           <not installed>
```

</details>

**Comment by Object905** [comment]
Well I happen to implement that 😃 

About `until the DataFrame is created` - is the bottleneck here is creation of LazyFrames?
Because LazyFrames are actually always created eagerly from python (in [__init__ ](https://github.com/pola-rs/polars/blob/54218e7e35e3defd4b0801e820c56eea6b91e525/py-polars/polars/lazyframe/frame.py#L308)it creates regular DataFrame by iterating over all objects, then calls .lazy()).
And later collect_async does only join+with_columns part asynchronously.
To support something like this LazyFrames has to support "scanning" python iterables/generators, etc, which would actually be a great feature.

It does use polars threadpool [here](https://github.com/pola-rs/polars/blob/54218e7e35e3defd4b0801e820c56eea6b91e525/crates/polars-python/src/lazyframe/general.rs#L595) that runs collect on background, and then only resolves the future with GIL acquired.
It should work with scan_parquet, scan_scv, scan_*, because they're executed on rust side in .collect and not eagerly.

**Comment by winstxnhdw** [comment]
> About until the DataFrame is created - is the bottleneck here is creation of LazyFrames?

Hmm.. Is there a way I can verify this? The `LazyFrame` only has ~50 rows but it is blocking the event loop for more than 10 seconds, which I think is an usually long time for creating such a trivial `DataFrame`.

> It does use polars threadpool [here](https://github.com/pola-rs/polars/blob/54218e7e35e3defd4b0801e820c56eea6b91e525/crates/polars-python/src/lazyframe/general.rs#L595) that runs collect on background, and then only resolves the future with GIL acquired.

I see that you are spawning the pool in Rust. I am not sure about PyO3 but in my experience, I found that spawning a separate background thread in C++ extensions would still block the event loop as Python still has to constantly poll the job for completion.

**Comment by Object905** [comment]
Well it's a different story then if it's only 50 rows. Care to provide data to reproduce?

Theoretically It shouldn't block event loop, because, because it's only resolving a future by calling this callback [here ](https://github.com/pola-rs/polars/blob/54218e7e35e3defd4b0801e820c56eea6b91e525/py-polars/polars/_utils/async_.py#L84) at the end of exectuon of .collect on rust side,
And does not hold GIL during actual collect.

**Comment by winstxnhdw** [comment]
I will get the data for you tomorrow.

> not hold GIL during actual collect

Not holding the GIL is one thing, but it still needs to run in a separate thread for it to not block the event loop. If resolving the future involves some kind of polling of the results, the event loop will still be blocked.

**Comment by Object905** [comment]
Internally it may pool on the results, but if it were to pool by blocking whole loop that would defeat the whole purpose of [Future](https://docs.python.org/3/library/asyncio-future.html#asyncio.Future)

**Comment by Object905** [comment]
> Hmm.. Is there a way I can verify this? (LazyFrame creation)

You can try to create LazyFrames on separate lines and time only that (or event just print-debug). Because [__init__ is sync](https://github.com/pola-rs/polars/blob/54218e7e35e3defd4b0801e820c56eea6b91e525/py-polars/polars/lazyframe/frame.py#L308).

**Comment by winstxnhdw** [comment]
You are right. LazyFrame's creation is indeed the one responsible for blocking the event loop. I took your advice and tried using `scan_ndjson` but it was still blocking the event loop.

```python
def get_intersection(transcription: IO[bytes], diarisation: IO[bytes]) -> Awaitable[DataFrame]:
    intersection_expression = min_horizontal(
        'end_time',
        'end_time_right',
    ) - max_horizontal(
        'start_time',
        'start_time_right',
    )

    return (
        scan_ndjson(transcription)
        .join(scan_ndjson(diarisation), how='cross')
        .with_columns(intersection_expression.alias('intersection'))
        .collect_async()
    )
```

**Comment by deanm0000** [comment]
When is the blocking happening? The output of the function is an awaitable so it seems like we need more info.

Are you doing something simple like this?
```python
df_awaitable= get_intersection(tr, di)
df = await df
```

more like this?
```python
asyncio.get_running_loop().run_until_complete(df_awaitable)
```

something else?

**Comment by winstxnhdw** [comment]
I am doing the former. I expect Awaitables to use the event loop. It defeats the purpose to use `collect_async` if I were to also run it in a thread pool. I could just use `collect` then.

**Comment by deanm0000** [comment]
You still have the issue that you're scanning an IO object so even though the polars binary would be non-blocking, it's getting its data from python which is blocking.

Try either making the inputs eager DFs or files. 

In the case where the input is `LazyFrame(transcription: dict)` you're going to have to wait for polars to parse through all its dicts to send to arrow memory.

In the case where the input is `scan_ndjson(transcription: IO)` every time scan is looking at data it needs python which (I think) would be blocking.

So maybe try one of these

```python
def get_intersection(transcription: pl.DataFrame, diarisation: pl.DataFrame) -> Awaitable[DataFrame]:
    transcription=transcription.lazy()
    diarisation=diarisation.lazy()
```
or
```python
def get_intersection(transcription: Path, diarisation: Path) -> Awaitable[DataFrame]:
    transcription=pl.scan_ndjson(transcription)
    diarisation=pl.scan_ndjson(diarisation)
```

**Comment by winstxnhdw** [comment]
The problem is that `polars` should be non-blocking end-to-end. This should be possible since reading from IO is async/await’s strength.

For now, `collect_async` is redundant in practice, and running `collect` in a thread pool + wrap futures would have the same effect. I suggest we add this to the documentation.

**Comment by Object905** [comment]
Remembered that there is https://github.com/pola-rs/polars/pull/17939, have not tried myself, but looks promising.
It doesn't support async and is more about releasing GIL for threading, but in future it may be possible to extend this approach to asyncio with https://github.com/PyO3/pyo3/issues/1632.
[pyo3-asyncio](https://github.com/awestlake87/pyo3-asyncio) seems to be plausible too, but it [lags](https://github.com/awestlake87/pyo3-asyncio/pull/125) in maintenance a bit.

Then it would be possible/nice to have something, like `scan_python` which will correspond to scanning python (async)generators, just like scan_csv,parquet,etc. does.
